File: src/Java/adventofcode2018/Day18.java
package adventofcode2018;

import java.util.Arrays;
import java.util.List;

import utils.Input;

/**
 * https://adventofcode.com/2018/day/18
 */
public class Day18 {

    public static void main(String[] args) {
        int[][] map = new Input(2018, 18).getMap();
        List<Character> states = Arrays.asList('.', '|', '#');
        int h = map.length, w = map[0].length;
        int max = 5000;
        int scores[] = new int[max];
        for (int t=0; t<max ; t++) {
            final int[][] mapCopy = Arrays.stream(map)
                    .map(int[]::clone)
                    .toArray(int[][]::new);
            for (int y=0; y<h; y++) {
                for (int x=0; x<map[y].length; x++) {
                    int fx = x, fy = y;
                    int counts[] = new int[3];
                    Arrays.stream(new int[][] {{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}})
                            .map(dir -> new int[] {fx+dir[0], fy+dir[1]})
                            .filter(dir -> dir[0] >= 0 && dir[0] < w && dir[1] >= 0 && dir[1] < h)
                            .forEach(dir -> counts[states.indexOf((char)mapCopy[dir[1]][dir[0]])]++);
                    if (mapCopy[y][x] == '.' && counts[1] >= 3) map[y][x] = '|';
                    if (mapCopy[y][x] == '|' && counts[2] >= 3) map[y][x] = '#';
                    if (mapCopy[y][x] == '#' && (counts[1] == 0 || counts[2] == 0)) map[y][x] = '.';
                }
            }
            int counts[] = new int[2];
            Arrays.stream(map).flatMapToInt(Arrays::stream).forEach(c -> {
                if (c == '|') counts[0]++;
                if (c == '#') counts[1]++;
            });
            scores[t] = counts[0] * counts[1];
        }
        System.err.println(scores[9]);
        for (int t2 = max-2; t2 > 0; t2--) {
            if (scores[t2] == scores[max-1]) {
                int period = max - 1 - t2;
                System.err.println(scores[1000_000_000 % period + 100 * period - 1]);
                break;
            }
        }
    }
}

File: src/Python/2018/Day18.py

from utils import Input, printResult
import copy

# https://adventofcode.com/2018/day/18

field = list(map(list, Input(2018, 18).lines()))
states = [".", "|", "#"]
h, w = len(field), len(field[0]) - 1
m = 1000
scores = [0] * m
for t in range(m):
    field2 = copy.deepcopy(field)
    for y in range(h):
        for x in range(w):
            counts = [0, 0, 0]
            for dir in [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]]:
                p = [x + dir[0], y + dir[1]]
                if p[0] >= 0 and p[0] < w and p[1] >= 0 and p[1] < h:
                    counts[states.index(field2[p[1]][p[0]])] += 1
            if field2[y][x] == '.' and counts[1] >= 3: field[y][x] = "|"
            if field2[y][x] == '|' and counts[2] >= 3: field[y][x] = "#"
            if field2[y][x] == '#' and (counts[1] == 0 or counts[2] == 0): field[y][x] = "."
    counts = [
        sum(line.count("|") for line in field),
        sum(line.count("#") for line in field)
    ]
    scores[t] = counts[0] * counts[1]

printResult(1, scores[9])

for t in range(m-2, 1, -1):
    if scores[t] == scores[m-1]:
        period = m - 1 - t
        printResult(2, scores[m + 1000_000_000 % period - m % period - period - 1])
        break
